pkg/core/runtime/component/finalizer.go

```diff
@@ -0,0 +1,40 @@
+package component
+
+import (
+	"sync"
+
+	"github.com/kumahq/kuma/pkg/util/channels"
+)
+
+// Finalizer is a helper for implementing GracefulComponent
+type Finalizer struct {
+	finishCh chan struct{}
+	mutex    sync.Mutex
+}
+
+func (f *Finalizer) Running() {
+	f.mutex.Lock()
+	defer f.mutex.Unlock()
+	if f.finishCh == nil || channels.IsClosed(f.finishCh) {
```

Comment:
Inconsequential most likely but I feel the channel being closed here is a programmer error.

Reply:
I agree, we can add a constructor `NewFinalizer` to always create a valid `&Finalizer{}` object

Reply:
finalizer is gone